### ISSUE TITLE
How to link system-provided shared object file?

### ISSUE THREAD

**Issue by wbrickner** [BrainiumLLC/cargo-mobile#12]
Hello,

I am targeting the `Oculus Quest 2`, and have prepared my own bindings via bindgen for their Mobile SDK.
I have gotten the library to link - the build succeeds.

At runtime (*on my android phone, and also on the physical Quest 2*), the app launches and immediately crashes, complaining that the primary application `my_app.so` generated by `cargo-mobile` depends on my `libvrapi.so`:

```
01-06 19:00:02.522  6664  6664 W NativeActivity: NativeActivity LoadNativeLibrary("/data/app/com.willbrickner.power_pilot-gLvP0sExXMtQZk8DzeK19g==/lib/arm64/libpower_pilot.so") failed: dlopen failed: library "libvrapi.so" not found
01-06 19:00:02.523  6664  6664 E AndroidRuntime: Process: com.willbrickner.power_pilot, PID: 6664
01-06 19:00:02.523  6664  6664 E AndroidRuntime: java.lang.UnsatisfiedLinkError: Unable to load native library "/data/app/com.willbrickner.power_pilot-gLvP0sExXMtQZk8DzeK19g==/lib/arm64/libpower_pilot.so": dlopen failed: library "libvrapi.so" not found
01-06 19:00:02.526   996  6584 W ActivityTaskManager:   Force finishing activity com.willbrickner.power_pilot/android.app.NativeActivity
```

This only partially makes sense, shouldn't the `libvrapi.so` be on the system?  Isn't that the entire point of having a shared library? To have it be shared among all the binaries which make use of it? I never gave any special instructions to include the `libvrapi.so` *literally in the APK bundle*, should I have? How would I do that?

Thank you, this project will allow so much technological progress!

**Comment by wbrickner** [comment]
I've figured out that I can call functions in `libvrapi.so` if it is placed in the `./gen/android/{{PROJECT_NAME}}/app/src/main/jniLibs/arm64-v8a` before building / installing, no runtime linker errors.

This worries me, are these folders ever going to be modified by `cargo-mobile`? Is there some way to automate this process? I'm new to using rust to call into other dynamic libraries.

Hopefully I can modify something like the `mobile.toml` to copy this file from the `ovr-mobile-sys` crate I have generated automatically, as I'd like to contribute my progress back as a template or something.

Thanks!

**Comment by francesca64** [comment]
> Thank you, this project will allow so much technological progress!

Thanks so much! I'm really glad you find it useful.

> This only partially makes sense, shouldn't the `libvrapi.so` be on the system?  Isn't that the entire point of having a shared library?

~~The point of shared libraries is actually to make linking a much more complicated problem~~

Do you have any documentation that `libvrapi.so` should be present on the system? It might seem like an odd question, but i.e. `libc++_shared.so`, despite its name, is _not_ system-wide and has to be bundled in jniLibs to be used.

From doing a search, it looks like it's something you're intended to bundle: https://github.com/search?p=1&q=libvrapi.so&type=Code

...but almost all of the results are in the UE4 source code, which I guess won't show up for you. I'm specifically seeing a build step that copies `libvrapi.so` into their equivalent of the jniLibs dir.

> This worries me, are these folders ever going to be modified by `cargo-mobile`?

That's where we put symlinks to your compiled app `.so`, and also `libc++_shared.so` if we detect that your app depends on it. cargo-mobile _will_ check if any symlinks present are valid, and [delete them if they're broken](https://github.com/BrainiumLLC/cargo-mobile/blob/3c6237334be0ff531678eaa98325498b357039da/src/android/jnilibs.rs#L82). It should therefore be safe to do this, though you'd want to change a `.gitignore` or two.

However, cargo-mobile is billed as being convenient, so if you did that, it'd ruin my reputation! We can do better.

> Hopefully I can modify something like the `mobile.toml` to copy this file from the `ovr-mobile-sys` crate I have generated automatically

I think I like the idea of using the `metadata` section in `Cargo.toml` for this, which is what we currently do for specifying iOS frameworks to include. It seems like a natural place to put build configuration, though I suppose the line between what goes in `mobile.toml` and `Cargo.toml` is currently poorly defined.

Would you like a design like this?

```toml
[package.metadata.cargo-android.libs]
libA = { armeabi-v7a = "/wherever/armeabi-v7a/libA.so", arm64-v8a = "/wherever/arm64-v8a/libA.so" }
libB = { armeabi-v7a = "/wherever/armeabi-v7a/libB.so", arm64-v8a = "/wherever/arm64-v8a/libB.so" }
```

cargo-mobile would automatically symlink the specified libs into jniLibs, so you wouldn't have to worry about anything. Though, is having a fixed path going to be difficult for your use-case?

> as I'd like to contribute my progress back as a template or something.

That'd be super cool! Your entire use-case is super cool, just so you know.

**Comment by wbrickner** [comment]
> Do you have any documentation that libvrapi.so should be present on the system?

Yes, I have a shell and I can see that `libvrapi.so` is in ` /system/lib/libvrapi.so`.

> Would you like a design like this?

Yes this looks very good! I see two drawbacks: 

- having to specify an absolute path, but it is no worse than having to place the file in a directory I guessed correctly, and it is good to abstract away from the big two's toolchains.
- I would not guess library configurations would be controlled by a field called `metadata`

> That'd be super cool! Your entire use-case is super cool, just so you know.

Thank you! I hope to be the first™ to get a production VR tool written in rust running on the Quest 2, or maybe in general I haven't checked.  I am having trouble unrelated to `cargo-mobile`, so hopefully I make it through. If you're interested, it's a VR experience that lets you plug in a DJI controller into the Quest 2's USB-C port and fly the drone in FPV using hand gestures rather than joysticks. Unfortunately after battling with the JVM, the JNI, the NDK, the OVR SDK, the Android System APIs, and a reputation-preservingly small amount of friction with `cargo mobile`, I have discovered that ~~Facebook~~ Oculus has blocked userspace applications from using these USB APIs, but I have a workaround in progress to get around this and if I release it I'll link back to `cargo-mobile`.

Leaving this issue open as a tracker for library config support in `mobile.toml`.

**Comment by wbrickner** [comment]
Thinking about this more carefully, do you think there'd be a way to discover that a rust dependency bundles a dynamic library, and automatically symlink into the `./target` directory? This would be really cool, and to the end user would look like magic. I know this is often done in `bindgen` crates, where the bindings are rust declarations, and then their `build.rs` automatically adds in things for cargo to link against, like:

```rust
println!("cargo:rustc-link-search={}/VrApi-1.40/Libs/Android/{}/Release", env!("CARGO_MANIFEST_DIR"), abi);
println!("cargo:rustc-link-lib=dylib=vrapi");
```

In a crate worth using this is also guarded by platform, could `cargo-mobile` inspect this and symlink the linked libraries? That would probably be the ideal solution, to have no configuration and make it seem like magic.

However, you could get more generality out of the manual configuration approach, because let's say I want to add in a gradle dependency, I would prefer to put that in `mobile.toml` than in some gradle config file.

**Comment by francesca64** [comment]
> Well, I have a shell and I can see that `libvrapi.so` is in `/system/lib/libvrapi.so`

Weird, then I share your expection that we shouldn't need to bundle it.

This might be a dumb question, but does it make a difference if you use `/system/lib` for your `rustc-link-search` path?

> having to specify an absolute path

What would be ideal here? Our existing convention is that relative paths are interpreted as relative to the project root, if that helps at all.

> I would not guess library configurations would be controlled by a field called `metadata`

Fair point! `package.metadata` is just `Cargo.toml`'s designated field for arbitrary data, though it's not very commonly used. Even ignoring intuitiveness, it's likely preferable to have all cargo-mobile config live under one roof (`mobile.toml`), right?

> If you're interested, it's a VR experience that lets you plug in a DJI controller into the Quest 2's USB-C port and fly the drone in FPV using hand gestures rather than joysticks.

I know I already said that your use-case is super cool, but that's still much cooler than I expected. Everyone at Brainium already has a Quest 2, so when you succeed my boss will have the perfect excuse to get a drone.

> I have discovered that ~~Facebook~~ Oculus has blocked userspace applications from using these USB APIs, but I have a workaround in progress and if I release it I'll link back to `cargo-mobile`.

Good luck, and thanks! I'm rooting for you.

> do you think there'd be a way to discover that a rust dependency bundles a dynamic library, and automatically symlink into the `./target` directory? This would be really cool, and to the end user would look like magic.

We already do some magic using `readelf` to detect if we need to symlink `libc++_shared.so`, so doing even more magic sounds fun!

As for whether or not it's viable... to read `build.rs` output, it sounds like [we'd have to pass `-vv` to cargo](https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script), but that doesn't help if the build script doesn't actually run. I suppose it's something we could cache, but optimistically, it could already be cached somewhere in `target` and/or be seen elsewhere in the build output. We don't have too much ability to do `cargo build` in interesting ways though, since that output still goes to the user. 

> However, you could get more generality out of the manual configuration approach, because let's say I want to add in a gradle dependency, I would prefer to put that in `mobile.toml` than in some gradle config file.

Yeah, having an explicit/manual option is always good. There's no reason not to have both, besides if the magic solution is impossible, naturally!
